File: scratch/CRCTest.java
import java.io.FileInputStream;
import java.io.File;

public class CRCTest {
    public static void main(String[] args) throws Exception {
        File file = new File("/Users/cfayaz/Documents/Backup_codes/Backup_2.key");
        byte[] keyfilePool = new byte[64];
        FileInputStream stream = new FileInputStream(file);
        
        long crc = 0xFFFFFFFFL;
        byte[] buffer = new byte[8192];
        int totalRead = 0;
        
        while (totalRead < 1048576) {
            int toRead = Math.min(8192, 1048576 - totalRead);
            int read = stream.read(buffer, 0, toRead);
            if (read <= 0) break;
            
            for (int i = 0; i < read; i++) {
                int byteVal = buffer[i] & 0xFF;
                crc = crc ^ byteVal;
                for (int j = 0; j < 8; j++) {
                    if ((crc & 1L) != 0L) {
                        crc = (crc >>> 1) ^ 0xEDB88320L;
                    } else {
                        crc = (crc >>> 1);
                    }
                }
                
                int poolIndex = (totalRead + i) % 64;
                long crcHighByte = (crc >>> 24) & 0xFFL;
                keyfilePool[poolIndex] = (byte)(keyfilePool[poolIndex] + crcHighByte);
            }
            totalRead += read;
        }
        stream.close();
        
        for (int i = 0; i < 64; i++) {
            System.out.printf("%02X", keyfilePool[i]);
        }
        System.out.println();
    }
}

File: scratch/crc_test.kt
import java.io.File

fun main() {
    val keyfiles = listOf(File("/Users/cfayaz/Documents/Backup_codes/Backup_2.key"))
    val keyfilePool = ByteArray(64)
    for (file in keyfiles) {
        val stream = file.inputStream()
        var crc = 0xFFFFFFFFL
        val buffer = ByteArray(8192)
        var totalRead = 0
        
        while (totalRead < 1048576) {
            val toRead = Math.min(8192, 1048576 - totalRead)
            val read = stream.read(buffer, 0, toRead)
            if (read <= 0) break
            
            for (i in 0 until read) {
                val byteVal = buffer[i].toInt() and 0xFF
                crc = crc xor byteVal.toLong()
                for (j in 0..7) {
                    crc = if ((crc and 1L) != 0L) {
                        (crc ushr 1) xor 0xEDB88320L
                    } else {
                        crc ushr 1
                    }
                }
                
                val poolIndex = (totalRead + i) % 64
                val crcHighByte = (crc ushr 24) and 0xFF
                keyfilePool[poolIndex] = (keyfilePool[poolIndex].toInt() + crcHighByte.toInt()).toByte()
            }
            totalRead += read
        }
        stream.close()
    }
    println(keyfilePool.joinToString("") { String.format("%02X", it) })
}
